### ISSUE TITLE
Add new DeserializationFeature, "FAIL_ON_MULTIPLE_ROOT_LEVEL_VALUES" or something of the sort

### ISSUE THREAD

**Issue by 2rs2ts** [FasterXML/jackson-databind#800]
#726 seems to be extremely troubling. Consider the following:

```
scala> import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.ObjectMapper

scala> val mapper = new ObjectMapper()
mapper: com.fasterxml.jackson.databind.ObjectMapper = com.fasterxml.jackson.databind.ObjectMapper@60a230a0

scala> mapper.readValue("""{}: "foo"}""", classOf[java.util.Map[Any, String]])
res1: java.util.Map[Any,String] = {}
```

There's nothing about `{}: "foo"}` that should make it be considered representing valid JSON, at least in my opinion.

Therefore, I think there should be a `DeserializationFeature` which will cause an exception to be thrown when encountering multiple root-level values in a given content string. I'm not sure if it should be a `JsonParseException` or a `JsonMappingException`.

**Comment by cowtowncoder** [comment]
If you do want to ensure that there is nothing following after valid value you may construct a `JsonParser` explicitly, call mapper.readValue(), and then advance parser to see that it gets `null` (expected), instead of a parsing exception or token. And in latter case, report error.
I think some users do this to ensure that the input only contains a single value. In your case it would result in a `JsonParseException`, but only if explicitly trying to read more from the underlying `JsonParser`.

But I think addition of a new feature could work, or, maybe an alternative: add a new method in `ObjectReader`.
Something like:

```java
MyPOJO pojo = mapper.readerFor(MyPOJO.class).readOnlyValue(input);
```

(or whatever would be better name)

which would basically check the input as described above.
One downside of this approach would be that a few new methods are needed.

So maybe `DeserializationFeature` actually makes more sense.

As to exception, sounds more like `JsonMappingException` to me, although I can see why `JsonParseException` would be appropriate (when following strict JSON spec definition of a single root value).

**Comment by 2rs2ts** [comment]
@cowtowncoder I'm actually biased towards new methods but `DeserializationFeature`s seem to be idiomatic. New methods, I think, are nicer because you know from an API standpoint what you will get rather than being dependent on the state of an `ObjectMapper` at runtime.

I'm not familiar with the databind source beyond what I've delved into while debugging my own code. Still, this is pretty important to my team (https://github.com/paypal/cascade/issues/105), so I was planning on taking a swing at implementing this. If you'd like me to do that, would be willing to explicitly define the requirements (new methods or `DeserializationFeature`, exception type) so I know what to shoot for, and could help me with the CLA (not familiar with this sort of thing), then I'd be glad to help.

**Comment by cowtowncoder** [comment]
@2rs2ts Sounds good to me!

I agree in that new methods would be more explicit, and I like that approach. It would also keep `readValues()` separate and not cause problems there. There are some trade-offs (maybe some users would prefer forcing this on all access), but all in I think it makes sense.

CLA can be found here: https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
and is one-page document that needs to be filled with simple contact info, printed, scanned and email to `info` at fasterxml dot com. We'll store it in a private repo to know not to ask for another copy for further contributions. :)

On implementation, I think that:

1. Name of method to use is up to you -- renaming is easy, before code gets released. So we just need to finalize it before final 2.6.0.
2. Exception... despite what I initially said, maybe `JsonParseException` makes more sense. I assume from caller perspective it is indeed JSON well-formedness problem -- there should not be further values; and whether it's just garbage (non-whitespace), or tokens, makes no difference
3. Handling should just call `JsonToken t = p.nextToken()`, and throw exception if `t` is not null, indicating type of token. Whether to include actual value (or substring thereof) I don't know -- it's probably useful, but may be problematic for huge input, or garbage.

One way to do this would be for you to work on the PR, and I can have a look, suggest improvements, and once things look good I can merge it. If minor tweaks are needed I can apply them after merge as well.

**Comment by 2rs2ts** [comment]
@cowtowncoder Got it, you'll likely hear from me within a week.

**Comment by 2rs2ts** [comment]
@cowtowncoder Sorry to have taken so long to reply. The contributor agreement is at odds with the IP agreement I signed as part of working at PayPal. I don't think I can contribute while I still work at PayPal.

**Comment by cowtowncoder** [comment]
95%+ of people sign individual CLA, and not corporate one, for what that is worth. Corporate CLA is only applicable if you write code as part of the job (that is, using company's systems, facilties, work time, or within domain you work at during your job).
But if this is as part of your job I understand.

The only other thing I'll add for sake of completeness is that almost all corporate CLAs are using standard Apache one, and not the one I linked to. I will really need to upload that one, since it was requested by lawyers from 2 big fortune-500 companies (a big social network company, and a huge e-tailer, no names mentioned :-D).
So that could help. If not, it is unfortunate, but not completely unheard of.

**Comment by cowtowncoder** [comment]
Here's the link to the Corporate CLA:

https://github.com/FasterXML/jackson/blob/contributor-agreement-corporate.txt

now included on Jackson project portal page at https://github.com/FasterXML/jackson.

As I said, this is the CLA that corporate lawyers have been content with, having had some concerns about the individual CLA. This document is based on Apache corporate license (AFAIK), whereas the invidual CLA is based on an earlier Sun Microsystems CLA.

**Comment by cowtowncoder** [comment]
Actually #1583 was implemented for 2.9.0 and I think this is the same ask. Closing.
